changelog.d/2970.change.1.rst

```diff
@@ -0,0 +1,13 @@
+Added **experimental** support for ``pyproject.toml`` configuration
+(as introduced by :pep:`621`). Configuration parameters not covered by
+standards are handled in the ``[tool.setuptools]`` sub-table.
+
+In the future, existing ``setup.cfg`` configuration
+may be automatically converted into the ``pyproject.toml`` equivalent before taking effect
+(as proposed in :issue:`1688`). Meanwhile users can use automated tools like
+:pypi:`ini2toml` to help in the transition.
+
+Please note that the legacy backend is not guaranteed to work with
+``pyproject.toml`` configuration.
```

Comment:
Can we instead make it a clean failure? One of the frustating properties of working with distutils/setuptools is that there are a lot of ways to get subtle failures / accidental behaviours, and it'd be nice to eliminate a potential class of issues here by simply making it an explicit failure.

We can aways relax things later, if we so choose, so it'd be nice to start with strict and clear boundaries.

Reply:
Thank you very much for the review @pradyunsg.

I have nothing against raising a proper exception here. Just please note that right now there are no intrinsic problems with using `pyproject.toml` configurations with the legacy backend, I just wanted to make explicit that this is not part of the supported interface.

That aside, do you have any recommendation on how to implement this? Right now the configuration files are being read inside `setuptools.dist:Distribution.parse_config_files`, which is unaware about which backend was chosen.

For the "non-legacy" backend, in this PR I managed to avoid the "blind execution" of the `setup.py` script, so we can have more control about how/when all these functions are called. However for the "legacy" backend I purposefully wanted to keep the execution as backwards compatible as possible, so all parsing related functions are invoked via `distutils.core`, which is not easily changeable...